Documentation/proposals/202212-scrape-config.md

```diff
@@ -0,0 +1,150 @@
+scrape configurations. With the new `ScrapeConfig` CRD, it would be possible to consolidate some of that logic, where
+the other `*Monitor` CRDs could be migrated so that they create a ScrapeConfig resource that would ultimately be used by
+the operator to generate scrape configuration.
+
+## Pitfalls of the current solution
+
+Using `additionalScrapeConfig` comes with drawbacks:
+
+* Teams have to build an infrastructure to add scrape rules in a centralized manner, which creates a bottleneck since a
+  single team becomes responsible for the configuration
+* There is no input validation, which can lead to an invalid prometheus configuration
+
+# Goals
+
+* Provide a way for users to self-service adding scrape targets
+* Consolidate the scrape configuration generation logic in a central point for other resources to use
+
+## Audience
+
+* Users who serve Prometheus as a service and want to have their customers autonomous in defining scrape configs
+* Users who want to manage scrape configs the same way as for services running within the Kubernetes cluster
+* Users who want a supported Kubernetes way of scraping targets outside the Kubernetes cluster
+
+# Non-Goals
+
+* This proposal doesn't aim at covering all the fields in
+  [`<scrape_config>`](https://prometheus.io/docs/prometheus/latest/configuration/configuration/#scrape_config).
+  Specifically, no service discovery other than `static_configs` or `file_sd_configs` should be implemented at first.
+* Refactoring of the other CRDs is not in scope for the first version
+
+# How
+
+As described by
+[@aulig in #2787](https://github.com/prometheus-operator/prometheus-operator/issues/2787#issuecomment-559776221), we
+will create a new ScrapeConfig CRD, this new CRD will act the same as the other CRDs and append scrape configurations to
+the configuration. Usage of ScrapeConfig doesn't exclude the use of the other CRDs, they are not mutually exclusive.
+`ScrapeConfig` will allow for any scraping configuration, while the other CRDs provide sane defaults. This will allow
+for isolated testing of the new `ScrapeConfig` CRD.
+
+```mermaid
+graph TD;
+ServiceMonitor --> prometheusConfig
+PodMonitor --> prometheusConfig
+ScrapeConfig --> prometheusConfig
+prometheusConfig
+```
+
+Using a pseudo custom resource definition, we should have the following:
+
+```yaml
+apiVersion: monitoring.coreos.com/v1alpha1
+kind: ScrapeConfig
+metadata:
+  name: my-scrape-config
+  namespace: system-monitoring
+  labels:
+    test: value
+spec:
+  staticConfigs:
+    - <staticConfig>[] # new resource
+  fileSDConfigs:
+    - <fileSDConfig>[] # new resource
+  httpSDConfigs:
+    - <httpSDConfig>[] # new resource
+  relabelConfigs:
+    - <relabelConfig>[] # https://github.com/prometheus-operator/prometheus-operator/blob/e4e27052f57040f073c6c1e4aedaecaaec77d170/pkg/apis/monitoring/v1/types.go#L1150
+  metricsPath: /metrics
+```
```

Comment:
I assume that you didn't want to list all the fields but we intend to support all the scrape config fields (apart from the SD mechanisms which are currently out of scope)?

Reply:
Eventually yes, I leave this to the implementation. For instance, on the `http_sd_config`, we might not have OAuth2.0 right away but it could be implemented at a later stage.

Reply:
Added more to the proposal in that regard

Reply:
:+1: